[14.0][ADD] sale_stock_prebook: Add process to prebook a sale order's stock

mt-software-de opened the pull request:
Add process to prebook/reserve a sale order's stock before confirming it.

Depends on
~~https://github.com/OCA/stock-logistics-warehouse/pull/1807~~ NOT NEEDED ANYMORE
~~https://github.com/OCA/wms/pull/769~~ MERGED

mt-software-de committed:
[ADD] sale_stock_prebook: Add process to prebook a sale order's stock before confirming it

mt-software-de commented:
cc @jbaudoux

rousseldenis commented:
@mt-software-de What's the difference with stock_reserve_sale ? 

https://github.com/OCA/stock-logistics-warehouse/pull/1788

mt-software-de commented:
> @mt-software-de What's the difference with stock_reserve_sale ? 
> 
> https://github.com/OCA/stock-logistics-warehouse/pull/1788

Please read the comment of @jbaudoux https://github.com/OCA/stock-logistics-warehouse/pull/1696#issuecom 
This explains it very well.

jbaudoux commented on sale_stock_prebook/README.rst:
```diff
@@ -0,0 +1,81 @@
+Authors
+~~~~~~~
+
+* MT Software
```
Can you keep original authors of the initial version:
* Camptocamp
* BCIM

jbaudoux commented on sale_stock_prebook/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "name": "sale_stock_prebook",
+    "summary": "Add process to prebook a sale order's stock before confirming it",
+    "version": "14.0.1.0.0",
+    "author": "MT Software, Odoo Community Association (OCA)",
```
```suggestion
    "author": "MT Software, Camptocamp, BCIM, Odoo Community Association (OCA)",
```

mt-software-de committed:
[ADD] sale_stock_prebook_stock_available_to_promise_release:
Take moves into account when computing the previous promised qty
Keep date_priority of Confirmed sale.orders which were reserved before

mt-software-de commented:
https://github.com/OCA/wms/pull/769 is needed for this PR.

jbaudoux reviewed (approved):
LGTM

jbaudoux commented:
> [OCA/wms#769](https://github.com/OCA/wms/pull/769) is needed for this PR.

Can you update this PR description with this?

rafaelbn reviewed (approved):
Functionally tested 👍🏼 

In my opinion README with the USAGE section should be improved with all how to / step by step. In other case, less reviewer, less user adopting it. 😄 ❤️

mt-software-de commented:
This one is now ready to merge, i removed the dependency to stock_helper and included the very specific logic to stop the proc run, within the module itself.

lmignon reviewed (changes requested):
It's possible to process the picking generated for the reservation. It should not be possible to set quantity done on the move line nor to validate the picking.

jbaudoux commented:
> It's possible to process the picking generated for the reservation. It should not be possible to set quantity done on the move line nor to validate the picking.

This line https://github.com/OCA/sale-workflow/pull/2624/files#diff-2a79dc3628d7cbd98f204791750fadc712f3e43c11341eabb2c722c7157eb258R12 should prevent it. How come you made the move ready ?

rousseldenis commented:
@mt-software-de @jbaudoux What's ths status of this ?

mt-software-de commented:
> @mt-software-de @jbaudoux What's ths status of this ?

Sorry i was not working on this. 
I still need to check why the test are failing. On my dev env they are green.

mt-software-de committed:
Fix Testing sale.order action_cancel
By default sale_cancel_confirm overwrites the method
When there is a addon which is doing something afterwards
this results in an error in testing

mt-software-de commented:
> @mt-software-de @jbaudoux What's ths status of this ?

I found the error. 
Its again a module which changes the default odoo behavior. `sale_cancel_confirm`.
I changed the github test flow. 
https://github.com/OCA/sale-workflow/pull/2624/commits/0b3da4545f83f601fe754b513db5c60e562a0dd5

Besides that this one is ready to merge. 
This is already quite sometime in production

mt-software-de commented:
> > It's possible to process the picking generated for the reservation. It should not be possible to set quantity done on the move line nor to validate the picking.
> 
> This line https://github.com/OCA/sale-workflow/pull/2624/files#diff-2a79dc3628d7cbd98f204791750fadc712f3e43c11341eabb2c722c7157eb258R12 should prevent it. How come you made the move ready ?

ping @lmignon

lmignon commented:
I've just made a request for a new repo to group together the addons relating to stock pre-booking. see https://github.com/OCA/repo-maintainer-conf/pull/73

lmignon committed:
[IMP] sale_stock_prebook: Do not allow to set quantity done
on reserved stock moves

mt-software-de commented:
@lmignon @yostashiro i added your suggestions.

@lmignon i added a separate commit, with your changes.

lmignon committed:
[IMP] sale_stock_prebook: speedup installation on existing DB

rousseldenis reviewed (approved):
Code review

rousseldenis commented:
@mt-software-de Ready ?

mt-software-de commented:
> @mt-software-de Ready ?

Yes but should we wait for the new repo?

mt-software-de commented:
Replaced by https://github.com/OCA/sale-prebook/pull/1
